### DISCUSSION TITLE
Support Multiple Channels

### DISCUSSION THREAD

**Discussion by kylemclaren** [microfeed/microfeed#69] [Ideas]
Currently (as I see it), microfeed supports only one channel which owns one or more items. So that would be analogous to one podcast (show) with one or more podcast episodes.

If you wanted to add another podcast (show) with it's own feed, you would need to run another microfeed instance.

Would be much better to host everything in one instance.

Is this on the roadmap?

**Comment by wenbinf** [comment]
For simplicity, currently one instance one feed. In practice, you can use different subdomains.

In the future, we may consider adding support for sub-feeds. For example, your podcast has 1200 episodes, and you might create multiple categories that have their own feeds.

**Comment by Giggiux** [comment]
I decided that I'm going to use microfeed in a project very soon, but I will need to implement multiple categories/feeds in my personal fork: is this something that you can be interested in merging into the main project, @wenbinf ? If yes, I would love to discuss how to structure the changes in order to be good enough to be merged!

**Reply by wenbinf** [reply]
For the time being, we prefer keeping microfeed simple and only support a single feed per instance. The main reason is that it's nontrivial to build multi-feed, in terms of 1) admin UI design, 2) database schema design, 3) url path design...

However, I would still encourage you to try it in your own fork to access what kind of effort it'll take for you to get a simple prototype - with a working prototype, i may be convinced that it's doable to have a clean implementation merged into the main branch.

Curious - what are your use cases? why do you need multiple feeds? there may be quick solutions that don't involve code changes.

**Reply by orawalters** [reply]
@wenbinf I've got use cases for this if you're interested. Namely, it'd be nice to be able to post to my podcast's website with blog posts that don't pollute the podcast RSS feed, or have seperate feeds for audio/video episodes if we decide to start doing this. A work around for me would be the ability to create multiple feeds that can filter based on post type. I also had the idea of creating a company website, so all my podcasts would be on the same place, but my plan to work around that is to maintain a separate database for this that's populated by crawling my own RSS feeds. Curious if either of these fit in with @Giggiux's use cases.

**Reply by Giggiux** [reply]
I will need to build a single page app with multiple "feeds" of audio/video/text content, that needs to be independently page-able (both using the json feed as api, and also for their rss feeds). 

As far as I can understand, now the different kind of items in microfeed are showed in the same feed, in chronological order, without any categorisation. I would need to either add a "custom" category to each item and then be able to add a query to the endpoint(s), or just have different channels.

The first option would need to use an extra column on the `items` table.
UI is just a new field in the add new item page.

The second option seems doable extending the usage of the current `channels` table to create new channels, then adding a foreign key column to the `items` table.
UI would need a little study, but doesn't seem anything impossible: easiest path would adding a "Add/Edit Channel" and "List channels" tabs, and a new dropdown in the new Item page.

**Reply by wenbinf** [reply]
> I will need to build a single page app with multiple "feeds" of audio/video/text content, that needs to be independently page-able (both using the json feed as api, and also for their rss feeds).

How frequently do you need to add/update items in these feeds? 

In terms of getting things done, there's a solution that could work right away:

* Setup multiple microfeed instances, e.g., one for video-only items (like video.mywebsite.com), one for audio-only items (like audio.mywebsite.com)...
* Aggregate multiple feeds into a single feed either via rss or json (hosted on www.mywebsite.com)

> As far as I can understand, now the different kind of items in microfeed are showed in the same feed, in chronological order, without any categorisation. I would need to either add a "custom" category to each item and then be able to add a query to the endpoint(s), or just have different channels.
> 
> The first option would need to use an extra column on the `items` table. UI is just a new field in the add new item page.
> 
> The second option seems doable extending the usage of the current `channels` table to create new channels, then adding a foreign key column to the `items` table. UI would need a little study, but doesn't seem anything impossible: easiest path would adding a "Add/Edit Channel" and "List channels" tabs, and a new dropdown in the new Item page.

In the end of the day, microfeed is a typical CRUD app. It's totally doable to write code and customize to fit your need.

The database schema was designed to support multiple feeds from the get go. To support multiple feeds, you'll need to support many-to-many relations between items and channels. One item could be included in multiple channels. And one channel could include multiple items. I don't think you want to add a foreign key column to either channel or item table. Instead, you'll need to create a separate table for channel-item relationships. 

The roadmap is clear. Now the problem is finding time & motivation to do it :) 

One concern is that if we make microfeed too complex and page request performance would degrade. Cloudflare doesn't give users much time budget for page requests: https://community.cloudflare.com/t/how-is-cpu-time-per-request-measured-in-cloudflare-workers/49964

**Reply by wenbinf** [reply]
> @wenbinf I've got use cases for this if you're interested. Namely, it'd be nice to be able to post to my podcast's website with blog posts that don't pollute the podcast RSS feed, or have seperate feeds for audio/video episodes if we decide to start doing this. A work around for me would be the ability to create multiple feeds that can filter based on post type. I also had the idea of creating a company website, so all my podcasts would be on the same place, but my plan to work around that is to maintain a separate database for this that's populated by crawling my own RSS feeds. Curious if either of these fit in with @Giggiux's use cases.

How about this?
* Setup multiple microfeed instances, e.g., one for video-only items (like video.mywebsite.com), one for audio-only items (like audio.mywebsite.com)...
* Aggregate multiple feeds into a single feed either via rss or json (hosted on [www.mywebsite.com](http://www.mywebsite.com/))

Tools like Zapier would make it easy to combine multiple rss/json feed into a single feed.
